File: Java/977_Squares_of_a_Sorted_Array.java
// Time O(n)
// Space O(n)
class Solution {
    public int[] sortedSquares(int[] nums) {
        int n = nums.length;
        int[] ret = new int[n];
        int left = 0, right = n -1;
        for(int i = n-1; i>=0; i--){
            int square;
            if(Math.abs(nums[left]) < Math.abs(nums[right]) ){
                square = nums[right];
                right--;
            }else{
                square = nums[left];
                left++;
            }
            ret[i] = square * square;
        }
        return ret;
    }
}

File: Python/977_Squares_of_a_Sorted_Array.py
class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        n = len(nums)
        result = [0] * n # result: [0,0,0,0,0]
        left = 0
        right = n -1
        # range(start, stop, step)
        # start ≦ i < stop
        for i in range(right, -1, -1):
            if abs(nums[left]) < abs(nums[right]):
                square = nums[right]
                right -= 1
            else:
                square = nums[left]
                left += 1
            result[i] = square * square
        return result
